monitoring/ping.go

```diff
@@ -0,0 +1,263 @@
+			self.Name, node.Name,
+			roundtripLatency.ValueAtQuantile(pingRoundtripQuantile),
+			pingRoundtripQuantile)
+
+		pingRoundtripPercentile := roundtripLatency.ValueAtQuantile(pingRoundtripQuantile)
+		if pingRoundtripPercentile >= pingRoundtripThreshold.Nanoseconds() {
+			log.Warningf("%s <-ping-> %s = slow ping RoundTrip detected. Value %dns over threshold %s (%dns)",
+				self.Name, node.Name, pingRoundtripPercentile,
+				pingRoundtripThreshold.String(), pingRoundtripThreshold.Nanoseconds())
+		} else {
+			log.Debugf("%s <-ping-> %s = ping RoundTrip okay. Value %dns within threshold %s (%dns)",
+				self.Name, node.Name, pingRoundtripPercentile,
+				pingRoundtripThreshold.String(), pingRoundtripThreshold.Nanoseconds())
+		}
+	}
+
+	return nil
+}
+
+// storePingInHDR is used to store ping RoundTrip values in HDR Histograms in memory
+func (c *pingChecker) storePingInHDR(pingroundtripLatency int64, node serf.Member) error {
+	s, exists := c.roundtripLatency.Get(node.Name)
+	if !exists {
+		c.roundtripLatency.Set(node.Name,
+			hdrhistogram.New(pingRoundtripMinimum.Nanoseconds(),
+				pingRoundtripMaximum.Nanoseconds(),
+				pingRoundtripSignificativeFigures),
+			statsTTLPeriod)
+		s, _ = c.roundtripLatency.Get(node.Name)
+	}
+
+	nodeLatencies, ok := s.(*hdrhistogram.Histogram)
+	if !ok {
+		return trace.BadParameter("couldn't parse roundtripLatency as HDRHistogram on %s", c.serfMemberName)
+	}
+
+	tmpSnapshot := nodeLatencies.Export()
+	if len(tmpSnapshot.Counts) >= slidingWindowSize {
+		// pop element at index 0 (oldest)
+		nodeLatencies.Reset()
+
+		lowerLimit := len(tmpSnapshot.Counts) - slidingWindowSize
+		if lowerLimit < 0 {
+			lowerLimit = 0
+		}
+		// using len(tmpSnapshot.Counts)-1 as the upper limit as another value
+		// will be added a few lines below
+		for i := lowerLimit; i < len(tmpSnapshot.Counts)-1; i++ {
+			nodeLatencies.RecordValue(tmpSnapshot.Counts[i])
+		}
+	}
+
+	err := nodeLatencies.RecordValue(pingroundtripLatency)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("%d recorded ping RoundTrip values for node %s",
+		nodeLatencies.TotalCount(), node.Name)
+
+	return nil
+}
+
+// calculateRTT calculates the RoundTrip time between two Serf Cluster members
+func calculateRTT(serfClient *serf.RPCClient, self serf.Member, node serf.Member) (int64, error) {
+	selfCoord, err := serfClient.GetCoordinate(self.Name)
+	if err != nil {
+		return 0, err
+	}
+	if selfCoord == nil {
+		return 0, trace.NotFound("self node %s coordinates not found", self.Name)
+	}
+
+	otherNodeCoord, err := serfClient.GetCoordinate(node.Name)
+	if err != nil {
+		return 0, trace.NotFound("error getting coordinates: %s -> %v", node.Name, err)
+	}
+	if otherNodeCoord == nil {
+		return 0, trace.NotFound("could not find a coordinate for node %s -> %v", node.Name, err)
+	}
+
+	return selfCoord.DistanceTo(otherNodeCoord).Nanoseconds(), nil
+}
+
+// setProbeStatus set the Probe according to status or raise an error if one is passed via arguments
```

Comment:
set**s** ... raise**s**
You don't need this method - it does not provide anything beyond what's already available with `NewProbeFromErr`.

Reply:
in a comment above both you and Roman told me to not use log.Error() everywhere and instead wrap it inside a dedicated function... that's why this function is here now.

Reply:
yes, but in this case you could simply rewrite top-level `Check` like this:
```go
func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
	// FIXME: #1 RttThreshold will become configurable in future
	// FIXME: #2 Send RttThreshold value to metrics

	err := c.check(ctx, r)
	if err != nil {
		log.Warn(err.Error())
		r.Add(NewProbeFromErr(c.Name(), "", err))
		return
	}
	r.Add(NewSuccessProbe(c.Name()))
}
```